---
title: "Report"
author: "Bello Ketty, Bertotti Michela and Boccolini Valerio"
output: html_document
date: '`r Sys.Date()`'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

In the present report, RNA-seq count data extracted from different cancer datasets from the Cancer Genome Atlas (TCGA) were analyzed. Data is split between tumor and control samples. In particular, data used for this study derived from **uterine corpus endometrial carcinoma cells**.

# Analysis

## 1. Load data

```{r data}
load("Uterine_corpus_endometrial_carcinoma.RData")
```

The loaded file includes:

-   `raw_counts_df`: contains the raw RNA-seq counts;

-   `c_anno_df`: contains sample names and conditions (case or control);

-   `r_anno_df`: contains the ENSEMBL genes ids, the length of the genes and the genes symbols.

## 2. Extract protein coding genes

It is preferable to focus the analysis on protein-coding genes only. To do so, it is used the `biomaRt` package. `biomaRt` provides tools for querying the ENSEMBL database to retrieve genomic annotations.

```{r biomart, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
library(biomaRt)
```

```{r PCG}
ensembl <- useEnsembl(biomart="ensembl",dataset="hsapiens_gene_ensembl", mirror = "useast")
attributes <- listAttributes(ensembl)
g_f <- getBM(attributes = c("ensembl_gene_id", "gene_biotype",
                            "external_gene_name"),mart = ensembl,
             filters = ("ensembl_gene_id"), 
             values = (r_anno_df$ensembl_gene_id))

g_coding <- g_f[which(g_f$gene_biotype == "protein_coding"),]
PC_r_anno <- r_anno_df[which(r_anno_df$ensembl_gene_id %in% g_coding$ensembl_gene_id),]
PC_raw_counts <- raw_counts_df[which(row.names(raw_counts_df) %in% g_coding$ensembl_gene_id),]
```

`g_f` contains ID, biotype and gene name of the genes in `r_anno_df`. Finally, all the dataframes were filtered to keep only the coding genes.

## 3. Differential Expression Analysis

**Differential Expression Analysis (DEA)** identifies genes that show statistically significant differences in expression levels between two or more conditions, corresponding to cases and control in this study. This analysis helps to:

-   Identify biomarkers or genes of interest involved in specific conditions;

-   Explore biological pathways and mechanisms;

-   Validate hypotheses in experimental biology and drug discovery.

Differential Expression Analysis is here performed using `edgeR` package.

```{r libraries, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
library("GenomicFeatures")
library("ggplot2")
library("stringr")
library("tidyverse")
library("edgeR")
library("pheatmap")
```

Firstly, a threshold was set for the raw count and the number of replicates.

```{r threshold, results='hide', message=FALSE, warning=FALSE}
count_thr <- 20 #minimum read count threshold
repl_thr <- 5 #minimum number of replicates threshold

filter_vec <- apply(raw_counts_df,1,
                    function(y)max(by(y,c_anno_df$condition,
                                      function(x) sum(x>=count_thr))))

#statistics for the filtering
table(filter_vec)
```

From the resulting table it is possible to notice that 16225 genes satisfy the threshold of reads and are present in 30 replicates of the same condition.

Subsequently the filters were applied to the dataframes in order to maintain only the genes that have at least 20 reads and 5 replicates. 
```{r dimension}
filter_counts_df <- raw_counts_df[filter_vec>=repl_thr,] 
filter_anno_df <- r_anno_df[rownames(filter_counts_df),]

#check dimension consistency
dim(filter_counts_df)
dim(filter_anno_df)
```

To check for uniformity across samples, it is useful to visualize the library size of each sample:

```{r barplot, echo=FALSE, fig.cap="**Fig. 1** Barplot of total number of reads for each sample of the filtered dataframe.", fig.align='center'}
size_df <- data.frame("sample"=colnames(filter_counts_df), 
                      "read_millions"=colSums(filter_counts_df)/1000000) 

ggplot(data=size_df,aes(sample,read_millions)) +
  geom_bar(stat="identity",fill="indianred2",colour="indianred2",width=0.7,alpha=0.7)+
  coord_flip()+
  theme_bw()
```

It might also be useful to visualize the number of reads for genes in each sample. From the following boxplot it is clear that data need to be normalized before proceeding with DEA.

```{r boxplot1, echo=FALSE, fig.cap="**Fig. 2** Boxplot of variability of reads for genes among samples (before normalization).", fig.align='center'}
long_counts_df <- gather(as.data.frame(filter_counts_df), key = "sample", value = "read_number")

ggplot(data=long_counts_df,aes(sample,read_number+1)) + 
  geom_boxplot(colour="cyan4",fill="cyan4",alpha=0.7) +
  theme_bw() +
  scale_y_log10() + theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

Firstly we created a `DGEList` object: it contains information about count data (`filter_counts_df`), conditions (`c_anno_df$condition`), and gene annotations (`filter_anno_df`) and it is used by `edgeR` for normalization.

Secondly, in order to apply intra- and inter-sample normalization, the function `calcNormFactors` was called while specifying the TMM (Trimmed Mean of M-values) method. In this way, the function calculates the scaling factor based on the library size.

Lastly, normalized counts were converted to counts-per-million (CPM).

```{r normalization}
#create a DGEList object
edge_l <- DGEList(counts=filter_counts_df, group=c_anno_df$condition, samples=c_anno_df, genes=filter_anno_df) 

#normalization
edge_n <- calcNormFactors(edge_l, method="TMM")

#create a cpm table
cpm_table <- as.data.frame(round(cpm(edge_n),2))
```

By looking at the following boxplot it is possible to assess the effect of normalization on data: the distributions of the single samples are now comparable; normalization has reduced bias in expression distributions.

```{r boxplot2, echo=FALSE, fig.cap="**Fig. 3** Boxplot of variability of reads for genes among samples (after normalization).", fig.align='center'}
long_cpm_df <- gather(cpm_table, key = "sample", value = "CPM")

ggplot(data=long_cpm_df,aes(sample,CPM+1)) +
  geom_boxplot(colour="orchid3",fill="orchid3",alpha=0.7)+
  theme_bw()+
  scale_y_log10() + theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

It is now possible to proceed with Differential Expression Analysis. An experimental design matrix is defined based on case and control conditions.

```{r des matr}
group <- edge_n$samples$group 
# define the experimental design matrix
design <- model.matrix(~0+group, data=edge_n$samples)
colnames(design) <- levels(edge_n$samples$group)
rownames(design) <- edge_n$samples$sample
```

The `estimateDisp` function estimates the dispersion parameters for our gene expression data. 
Using `glmQLFit` a statistical model is fitted to our data using a quasi-likelihood approach.

```{r dispersion}
edge_d <- estimateDisp(edge_n, design)
edge_f <- glmQLFit(edge_d, design)
```

With `contro` the comparison between case and control was defined.

```{r contro}
contro <- makeContrasts("case - control", levels = design)
```

The function `glmQLFTest` was called to perform the statistical test for differential expression following the given comparison (`contro`). The result was sorted by the logarithm of the fold change and a cutoff on the p-value was applied.

```{r fit}
#Fit the model with generalized linear models
edge_t <- glmQLFTest(edge_f,contrast=contro)
DEGs <- as.data.frame(topTags(edge_t,n=20000,p.value = 0.01, sort.by = "logFC"))
```

For keeping track of the fold change of each gene, a new column ("class") was added to the dataframe, containing either "+", "-" or "=". DEGs were then filtered for significance (FDR \< 0.01) and categorized into:

-   "+": Up-regulated (logFC \> 1.5).

-   "-": Down-regulated (logFC \< -1.5).

-   "=": Not significantly different

```{r DEGs}
DEGs$class <- "=" 
DEGs$class[which(DEGs$logCPM > 1 & DEGs$logFC > 1.5 & DEGs$FDR < 0.01)] <- "+"  
DEGs$class[which(DEGs$logCPM > 1 & DEGs$logFC < (-1.5) & DEGs$FDR < 0.01)] <- "-"  

# Sort the DEGs by log fold-change in descending order
DEGs <- DEGs[order(DEGs$logFC,decreasing = T),]
```

In the following table are reported the counts of up-regulated, down-regulated, and neutral genes.

```{r table DEGs, echo=FALSE}
table(DEGs$class)
```

The results of the Differential Expression Analysis underlined the down-regulation of 1394 genes and the up-regulation of 1418 genes, while the remaining 9855 genes are not subjected to differential expression in this case.

A **volcano plot** is commonly used to display data from Differential Expression Analysis because it effectively combines statistical significance and biological relevance in a single visualization. On the x-axis we find the log-fold change while on the y axis the p-value; the vertical dashed lines represent the fold-change thresholds (logFC \> 1.5 or \< -1.5). The most significant DEGs are represented by the colored dots: green for the down-regulated genes and red for the up-regulated ones respectively.

```{r volcano, echo=FALSE, fig.cap="**Fig. 4** Volcano plot of DEA.", fig.align='center'}
DEGs$neg_log10_FDR <- -log10(DEGs$FDR)

volcano_plot <- ggplot(DEGs, aes(x = logFC, y = neg_log10_FDR, color = class)) +
  geom_point(alpha = 0.6, size = 1.5) +  # Points with transparency and size
  scale_color_manual(values = c("=" = "grey", "+" = "tomato3", "-" = "forestgreen")) + 
  labs(
    title = "Volcano Plot of Differential Expression",
    x = "Log2 Fold Change",
    y = "-Log10 Adjusted P-Value (FDR)"
  ) +
  theme_minimal() +
  theme(
    legend.position = "top",
    plot.title = element_text(hjust = 0.5, size = 16),
    axis.title = element_text(size = 14)
  ) +
  geom_hline(yintercept = -log10(0.01), linetype = "dashed", color = "black") +  # FDR threshold line
  geom_vline(xintercept = c(-1.5, 1.5), linetype = "dashed", color = "black")    # Fold-change thresholds

print(volcano_plot)
```

**Heatmaps** are essential for visualizing and interpreting expression data in differential analysis. Heatmaps display expression levels of genes across multiple samples. Hierarchical clustering on columns is exploited to group together samples based on their similarity in expressed genotype; meanwhile row gene clustering referes to genes that are co-expressed. The data was filtered to focus only on up- and down- regulated genes (differentially expressed genes).

```{r heatmap, echo=FALSE, fig.cap="**Fig. 5** Heatmap of DEA.", fig.align='center'}
cols <- cols <- c(ifelse(c_anno_df$condition == "case", "sienna2", "goldenrod1")) 
pal <- c("forestgreen","white","tomato3") 
pal <- colorRampPalette(pal)(100)

heatmap(as.matrix(cpm_table[which(rownames(cpm_table)%in%DEGs$ensembl_gene_id[which(DEGs$class!="=")]),]),
        ColSideColors = cols,cexCol = 0.5,margins = c(4,4),col=pal,cexRow = 0.2)
```

The clustering process managed to separate almost perfectly the cases (red) from the controls (yellow).

```{r text file, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
up_DEGs <- DEGs[which(DEGs$class=="+"),]
down_DEGs <- DEGs[which(DEGs$class=="-"),]

write.table(up_DEGs,file="up_DEGs.txt",row.names=F,col.names=T,sep="\t",quote=F)
write.table(down_DEGs,file="down_DEGs.txt",row.names=F,col.names=T,sep="\t",quote=F)
write.table(DEGs,file="DEGs.txt",row.names=F,col.names=T,sep="\t",quote=F)
```

## 4. Gene Set Enrichment Analysis (GSEA)

**Gene Set Enrichment Analysis (GSEA)** is a statistical technique used to determine whether a predefined set of genes shows statistically significance, coordinated differences in expression under two different conditions such as "treatment" vs. "control" or "healthy" vs. "disease". Instead of evaluating individual genes separately, GSEA assesses entire groups of genes that are functionally related, such as those involved in a particular pathway or biological process.

Gene Set Enrichment Analysis is here done using the `clusterProfiler` package.

```{r libGSEA, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
library(fgsea)
library(org.Hs.eg.db)
library(biomaRt)
library(clusterProfiler)
library(enrichplot)
library(ggnewscale)
library(DOSE)
library(pathview)
library(tidyverse)
```

Firstly, `entrezgene_id` and `external_gene_name` were retrieved, using the `biomaRt` package, for each gene in the DEGs dataset. In `convert` these mappings will be stored and with `merge()` the DEGs data is aligned to mappings in `ensembl_gene_id`.

```{r res_DEGs}
# Load results of DEG
DEGs <- read.table("DEGs.txt",header=T,sep="\t",as.is=T)

ensembl <- useEnsembl(biomart = "ensembl",dataset = "hsapiens_gene_ensembl", mirror = "useast")
convert <- getBM(attributes=c("ensembl_gene_id","entrezgene_id","external_gene_name"),
                 filters=c("ensembl_gene_id"), 
                 values=DEGs$ensembl_gene_id,
                 mart = ensembl)

DEGs <- merge(DEGs,convert)
DEGs <- DEGs[which(!is.na(DEGs$entrezgene_id)),]
DEGs <- DEGs[-which(duplicated(DEGs$entrezgene_id)),]
```

Up- and down- regulated genes are now stored in two different lists: `up_DEGs` and `down_DEGs`.

```{r up_down}
up_DEGs <- merge(up_DEGs,convert)
up_DEGs <- up_DEGs[which(!is.na(up_DEGs$entrezgene_id)),]
up_DEGs <- up_DEGs[-which(duplicated(up_DEGs$entrezgene_id)),]
down_DEGs <- merge(down_DEGs,convert)
down_DEGs <- down_DEGs[which(!is.na(down_DEGs$entrezgene_id)),]
down_DEGs <- down_DEGs[-which(duplicated(down_DEGs$entrezgene_id)),]
```

```{r ranks, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
## Create vector with ranks
ranks <- DEGs$logFC
ranks <- sort(ranks,decreasing = T)
names(ranks) <- DEGs$entrezgene_id 
head(ranks)
```

Now gene set enrichment analysis for **Gene Ontology (GO)** within the biological process (`ont = "BP"`) is performed. GO enrichment identifies which terms are statistically overrepresented (enriched) in the given set of genes, suggesting that those genes are involved in specific biological processes. The function `enrichGO` is used to perform Gene Set Enrichment Analysis on the list of up- and down- regulated genes and the human database (`OrgDb = org.Hs.eg.db`). The p-value need to be adjusted for multiple hypothesis testing, which was performed through the Benjamini-Hochberg (`pAdjustMethod = "BH"`) procedure. The outputs `ego_BP_up` and `ego_BP_down` will contain the enriched GO terms for biological processes associated with the genes. The results highlight which biological processes are overrepresented among up- or down-regulated genes, suggesting their involvement in specific pathways or responses.

```{r GO_BP}
#GO for BP on upregulated genes
ego_BP_up <- enrichGO(gene = up_DEGs$external_gene_name,
                      OrgDb = org.Hs.eg.db,
                      keyType = 'SYMBOL',
                      ont = "BP",
                      pAdjustMethod = "BH",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)

#GO for BP on downregulated genes
ego_BP_down <- enrichGO(gene = down_DEGs$external_gene_name,
                        OrgDb = org.Hs.eg.db,
                        keyType = 'SYMBOL',
                        ont = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05)
```

```{r vis_GO_up, echo=FALSE, fig.cap="**Fig. 6** Top 10 enriched up-regulated GO terms (BP)", fig.align='center'}
barplot(ego_BP_up,showCategory=10, main= "Top 10 enriched upregulated genes")
```

The up-regulated genes are strongly associated with processes involved in cell division, chromosome organization, and mitosis. Enrichment related to mitosis and chromosome segregation implies that the up-regulated genes are driving cell proliferation, this could lead to rapid tumor cells proliferation.

```{r vis_GO_down, echo=FALSE, fig.cap="**Fig. 7** Top 10 enriched down-regulated GO terms (BP)", fig.align='center'}
barplot(ego_BP_down,showCategory=10, main= "Top 10 enriched downregulated genes")
```

The top enriched GO terms for the down-regulated genes indicate that processes related to muscle contraction, heart function, and extracellular matrix organization are suppressed. The down-regulation of these biological processes likely reflects a combination of cancer progression, systemic effects like cachexia, and disruptions in tissue structure and function. These changes are hallmarks of advanced cancer and are associated with poor outcomes such as muscle wasting, cardiac dysfunction, and metastatic potential.

The same function is then applied to analyze the ontology related to molecular function (`ont = "MF"`), simply by changing the ontology parameter. Results of this functin are represented in the following barplots.

```{r GO_MF, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
#GO for MF on upregulated genes
ego_MF_up <- enrichGO(gene = up_DEGs$external_gene_name,
                      OrgDb = org.Hs.eg.db,
                      keyType = 'SYMBOL',
                      ont = "MF",
                      pAdjustMethod = "BH",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)

#GO for MF on downregulated genes
ego_MF_down <- enrichGO(gene = down_DEGs$external_gene_name,
                        OrgDb = org.Hs.eg.db,
                        keyType = 'SYMBOL',
                        ont = "MF",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05)
```

```{r vis_GO_MF_up, echo=FALSE, fig.cap="**Fig. 8** Top 10 enriched upregulated GO terms (MF)", fig.align='center'}
barplot(ego_MF_up,showCategory=10, main= "Top 10 enriched upregulated genes")
```

The top enriched up-regulated Molecular Functions (MF) reflect processes crucial for cancer progression, including cell division (microtubule binding and tubulin binding genes), cytoskeletal reorganization, cell adhesion changes (altered adhesion and peptidase regulation), and protease regulation. These biological processes support tumor growth, migration, and metastasis, which are hallmarks of aggressive cancers.

```{r vis_GO_MF_down, echo=FALSE, fig.cap="**Fig. 9** Top 10 enriched downregulated GO terms (MF)", fig.align='center'}
barplot(ego_MF_down,showCategory=10, main= "Top 10 enriched downregulated genes")
```

The down-regulation of these molecular functions is consistent with: ECM remodeling, loss of adhesion, altered signaling, increased metastasis, tumor progression, microenvironment alterations. This functional down-regulation aligns with hallmarks of cancer such as tissue invasion, metastasis, and microenvironmental deregulation.

**WikiPathways (WP) analysis**, is a type of pathway enrichment analysis that uses pathways from the WikiPathways database. WikiPathways is an open-access, community-driven database where scientists can collaboratively curate and maintain biological pathways across different species. It provides a wide range of pathways, including metabolic, signaling, regulatory, and disease pathways, much like KEGG or Reactome but with the added advantage of being open and editable.

WP analysis was performed using `enrichWP` function. Results of the analysis are reported below:

```{r WP}
#upregulated genes
eWP_up <- enrichWP(gene = up_DEGs$entrezgene_id,
                   organism = 'Homo sapiens',
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.1)

head(eWP_up, n=10)

#downregulated genes
eWP_down <- enrichWP(gene = down_DEGs$entrezgene_id,
                     organism = 'Homo sapiens',
                     pvalueCutoff = 0.05,
                     qvalueCutoff = 0.1)

head(eWP_up, n=10)
```

## 5. Visualization of one enriched pathway

Using KEGG enrichment analysis the top ten enriched pathways from up-regulated genes were visualized.

```{r KEGG, results='hide', message=FALSE, warning=FALSE}
#KEGG enrichment analysis
eWP_KEGGS <- enrichKEGG(gene = up_DEGs$entrezgene_id,
                        organism = 'human',
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.1)

head(eWP_KEGGS, n=20)
```

Here are reported the top 10 enriched KEGG pathways resulting from the up-regulated list of genes:

```{r KEGG_10}
knitr::kable(head(eWP_KEGGS[, 1:6], n=10))
```

The most enriched pathway is related to cell cycle (`ID: hsa04110`). It is possible to visualize it by using the package `pathview`:

```{r pathview, results='hide', message=FALSE, warning=FALSE}
logFC <- up_DEGs$logFC
names(logFC) <- up_DEGs$entrezgene_id 
pathview(gene.data = logFC, 
         pathway.id = "hsa04110", 
         species = "human")
```

Here is the network representation, with up-regulated genes represented in red.

```{r, echo=FALSE, out.width="80%", fig.cap="**Fig. 10** hsa04110 pathway", fig.align='center'}
knitr::include_graphics("hsa04110.pathview.png")
```

## 6. Identify enriched TFs

Now we focus on identifying **transcription factors (TFs) enriched** in the promoter sequences of up-regulated genes using motif enrichment analysis. Motif enrichment analysis determines whether specific TF binding motifs are statistically over-represented (enriched) in promoters of up-regulated genes. Over-representation indicates that certain TFs may play regulatory roles in the observed changes in gene expression. The used tool `PWMEnrich` use Position Weight Matrices (PWMs) to scan sequences and calculate enrichment scores. PWMs represent the likelihood of observing a specific nucleotide at each position of a binding site motif. Scoring and enrichment tests provide statistical significance values (p-values) for over-represented motifs.

```{r libTF, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
library(biomaRt)
library(MotifDb)
library(seqLogo)
library(PWMEnrich)
library(PWMEnrich.Hsapiens.background)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(BSgenome.Hsapiens.UCSC.hg38)
```

Firstly, with `getSeq()` we retrieve the actual DNA sequences of these promoters from the genome.

```{r TFs, message=FALSE, warning=FALSE}
genes <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene)
ensembl <- useEnsembl("ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
df <- getBM(attributes = c("external_gene_name",'entrezgene_id'),
            values=names(genes),filters ='entrezgene_id', mart = ensembl)
matched_indices <- match(genes$gene_id, df$entrezgene_id)
names(genes) <- ifelse(!is.na(matched_indices), df$external_gene_name[matched_indices], NA)

genes_name <- up_DEGs$external_gene_name # List of up-regulated genes
intersect_names <- intersect(names(genes), genes_name) #Filters genes to retain only those in both the up-regulated list and the UCSC database.

x <- promoters(genes,upstream = 500,downstream = 0)[intersect_names[intersect_names!= ""]] #Extracts promoter regions 500 bp upstream of the transcription start site (TSS).
seq <- getSeq(BSgenome.Hsapiens.UCSC.hg38,x)  
```

```{r lib_MES, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
# Calculating motif enrichment scores
library(MotifDb)   #Provides access to a curated collection of DNA sequence motifs.
library(seqLogo)   #Used for creating sequence logo plots.
library(PWMEnrich) #Offers functions for motif enrichment analysis.
library(PWMEnrich.Hsapiens.background)
```

Secondly, `PWMLogn.hg19.MotifDb.Hsap` loads a precomputed background model for motif enrichment using TF motifs from `MotifDb`. `motifEnrichment()` scans the promoter sequences (`seq`) using the background model and calculates motif enrichment scores using "affinity" scoring.

```{r MES, results='hide', message=FALSE, warning=FALSE}
data(PWMLogn.hg19.MotifDb.Hsap)
res = motifEnrichment(seq,PWMLogn.hg19.MotifDb.Hsap,score = "affinity")
```

```{r report}
report = sequenceReport(res, 1)
report
```

In the following plot are given the top 5 enriched TFs.

```{r plotMES, echo=FALSE, fig.cap="**Fig. 11** Top 5 enriched TFs.", fig.align='center'}
plot(report[1:5])
```

## 7. Motif analysis for the top enriched TFs

```{r lib_Man, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
library(MotifDb)
library(seqLogo)
library(PWMEnrich)
library(PWMEnrich.Hsapiens.background)
```

Next, it was analyzed the top enriched TF: RFX3. The retrieved motifs are then converted into Position Weight Matrices (PWMs) using the `toPWM` function. The `motifEcdf` function computes the empirical cumulative distribution function for the binding scores of each PWM. For each PWM's eCDF, the code calculates the 99.75th percentile cutoff. This is done using the `quantile` function applied to the eCDF object. The threshold is transformed into log2 scale for interpretability and appended to the thresholds list.

In the end, the calculated thresholds for each PWM are displayed. These thresholds can be used as binding score cutoffs for identifying regions of high TF binding affinity.

```{r toPWMs, message=FALSE, warning=FALSE, results='hide'}
mdb.human.selected = subset(MotifDb, organism=='Hsapiens' & geneSymbol=="RFX3")
if (length(mdb.human.selected) == 0) stop("No motifs found for the selected TF.")

motifs = as.list(mdb.human.selected)
length((mdb.human.selected))

pwms <- lapply(motifs, toPWM)
if (length(pwms) == 0) stop("No PWMs were generated for the selected TF.")

ecdf = motifEcdf(pwms,organism = "hg19",quick=TRUE)

# Initialize an empty list to store threshold values
thresholds <- numeric()
# Loop through each PWM's eCDF and calculate the threshold
for (ecdf_obj in ecdf) {
  # Calculate the 99.75 percentile threshold from the eCDF
  threshold <- log2(quantile(ecdf_obj, 0.9975))
  thresholds <- c(thresholds, threshold)  # Append to thresholds
}

# Display thresholds
threshold_data <- data.frame(TF = "RFX3", PWM = names(pwms), Threshold = thresholds)
print(threshold_data)
```
This means that there are 9 different documented binding motifs for RFX3 in the `MotifDb` database for humans. 
Having multiple motifs for the same transcription factor (RFX3) can mean several things:

-   Different Studies/Sources

-   Binding flexibility

-   Context-dependent

## 8. Pattern Matching

The following code identifies specific up-regulated genes whose promoter regions exhibit significant binding scores for RFX3 transcription factor motifs. `MotifScores` computes the binding scores of the provided transcription factor motifs (PWMs) for a set of promoter input sequences (`seq`). The output `scores` will be a matrix where each row corresponds to a DNA sequence, and each column represents a PWM. Each value indicates whether the score for a motif exceeds its threshold. `genes_id` will only contain a subset of `up_DEGs` with significant TF motif binding in their promoter regions.

```{r patMatch}
scores = motifScores(seq, pwms, raw.scores = FALSE, verbose = FALSE, cutoff = thresholds)
highscore_seq <- which(apply(scores,1,sum)>0)
genes_id <- up_DEGs[highscore_seq,]

dim(genes_id)
genes_id <- sort_by(genes_id, genes_id$PValue )
head(genes_id)
```
The resulting dataframe contains more than one thousand up-regulated genes that seems to be controlled by RFX3 TF.

## 9. PPI analysis

This section of the code focuses on identifying protein-protein interactions (PPI) among a set of differentially expressed genes (DEGs). The analysis leverages the STRING database to extract interaction data and export the resulting network in TSV format for further processing.

```{r lib_PPI, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
library(igraph)
library(tidyr)
library(biomaRt)
up_DEGs <- sort_by(up_DEGs, up_DEGs$PValue)[1:200,]
```

Firstly, were extracted the unique names of up-regulated DEGs and save them to a tab-separated file (`PPI_up_DEGs.txt`) for input into STRING. Secondly, were load STRING interaction data (`string_interactions.tsv`) containing PPI relationships between genes or proteins. Lastly, the previously exported DEG list was imported for integration with the interaction data.

```{r}
write.table(unique(up_DEGs$external_gene_name),sep = '\t', file = 'PPI_up_DEGs.txt',row.names = F, 
            col.names = F, quote = T)

links <- read.delim("string_interactions_short.tsv") # TSV data downloaded from STRING webpage 
UP__DEGs <- read.table("PPI_up_DEGs.txt", sep = "\t", header = TRUE)
```

## 10. Network

```{r lib_net, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
## Create nodes annotations using biomaRt
library(biomaRt)
library(igraph)
library(dplyr)
```

This section aims to create, annotate, and visualize a Protein-Protein Interaction (PPI) network using interaction data (`links`) and annotations retrieved via `biomaRt`. The workflow includes node annotation, graph creation, and customization of visual attributes to enhance biological insights.

```{r net}
ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = 'useast')

nodes <- getBM(attributes=c("external_gene_name","ensembl_gene_id","description","gene_biotype","start_position","end_position","chromosome_name","strand"),
        filters=c("external_gene_name"), 
        values=UP__DEGs [,1],
        mart = ensembl)

# Keep only relevant columns and remove duplicates
nodes <- unique(nodes[, c("external_gene_name", "description", "gene_biotype", 
                          "start_position", "end_position", "chromosome_name", "strand")])

valid_nodes <- unique(c(links$X.node1, links$node2))  

# Remove duplicate entries in the nodes data frame
# Ensure external_gene_name is unique
nodes <- nodes %>%
  distinct(external_gene_name, .keep_all = TRUE)

# Build the graph after the duplication
net <- graph_from_data_frame(
  d = links[, c("X.node1", "node2")],  
  vertices = nodes,                   
  directed = FALSE
)

net

## Plot the PPI network
plot(net, 
     edge.width=2,
     vertex.color="orange",
     vertex.size=10,
     vertex.frame.color="darkgray",
     vertex.label.color="black", 
     vertex.label.cex=0.7,
     edge.curved=0.1) 

# Find the node with the highest degree (most connections)
most_connected_node_23 <- V(net)[which.max(degree(net))]

# Create subgraph of this node and its immediate neighbors
neighborhood_net_23 <- make_ego_graph(net, order = 1, nodes = most_connected_node_23)[[1]]
plot(neighborhood_net_23)
```
